### ISSUE TITLE
Define variable to match declared variable with 'new'

### ISSUE THREAD

**Issue by NN---** [Microsoft/TypeScript#2149]
It is possible to define variable with declared when the function name is not 'new'

```typescript
interface X {
}
declare var X: {
    f(): X;
}

// OK, compatible type
var X = {
	f() { return <X>null; }
};


interface Y {
}
declare var Y: {
    new(): Y;
}

// Not compatible types
// new() : Y vs. new() => Y
var Y = {
	new() { return <Y>null; }
};
```

**Comment by ahejlsberg** [comment]
Yes, you can do that by putting the property name in quotes in the declaration:
```typescript
interface Y {
}
declare var Y: {
    "new"(): Y;  // Quotes to indicate this is not a construct signature
}

// OK, compatible type
var Y = {
    new() { return <Y>null; }
};
```

**Comment by NN---** [comment]
Oh, I didn't knew about quotes syntax.
Actually it is logical since identifier can be quoted, but most people don't do it.

Found it in the spec.

TS spec:
```
Method Signature:
  PropertyName ?opt CallSignature
```

ES spec:
```
PropertyName :
  IdentifierName
  StringLiteral
  NumericLiteral
```

It would be nice to mention 'new' case in the spec or in Q&A page.
Thanks.

**Comment by NN---** [comment]
@ahejlsberg The only problem is that I cannot change the declaration :)
It is in lib.d.ts
I am trying to define variable of Blob:

```typescript
interface Blob {
    type: string;
    size: number;
    msDetachStream(): any;
    slice(start?: number, end?: number, contentType?: string): Blob;
    msClose(): void;
}
declare var Blob: {
    prototype: Blob;
    new (blobParts?: any[], options?: BlobPropertyBag): Blob;
}
```

**Comment by ahejlsberg** [comment]
Not quite sure what you mean by "trying to define variable of Blob". Do you mean define a value that can be assigned to Blob? The Blob variable represents a constructor function (i.e. it is a variable of a constructor function type). To produce an assignment compatible value, you'll likely want to declare a class. For example:
```typescript
class MyBlob implements Blob {
    type: string;
    size: number;
    constructor(blobParts?: any[], options?: BlobPropertyBag) {
    }
    msDetachStream(): any {
        return undefined;
    }
    slice(start?: number, end?: number, contentType?: string): Blob {
        return undefined;
    }
    msClose(): void {
    }
}
```
You can then assign the class constructor function to the `Blob` variable:
```typescript
Blob = MyBlob;
```
Note, it is important to understand that there are two distinct and different things called Blob: A type (produced by the interface declaration) and a variable (with an *unnamed* constructor function type whose return value is the Blob interface). In type positions, Blob refers to the interface type, and in value positions (such as expressions) Blob refers to the constructor function.

**Comment by NN---** [comment]
I want to define the variable that is declared by 'declare var Blob : { ... }' .

```typescript
var Blob = {  
 // prototype and the constructor
};
```

**Comment by ahejlsberg** [comment]
You can't do that with an object literal because an object literal isn't "callable", i.e. it can't have a call or construct signature. Instead you need to write
```typescript
var Blob = MyBlob;
```
where `MyBlob` is a constructor function name introduced by a suitable class declaration (like my example above). Or you can assign some suitable constructor function object typed as `any` that you know has the correct shape.

It's important to understand that
```typescript
declare var Blob: {
    prototype: Blob;
    new (blobParts?: any[], options?: BlobPropertyBag): Blob;
}
```
doesn't introduce a property named "new", but rather introduces a construct signature that defines how to call the constructor function using the new operator, as in
```
var x = new Blob(...);
```
So, basically the declaration says that `Blob` is a variable of a constructor function type that (a) has a property named `prototype` and (b) can be called using `new` with the specified parameters.

**Comment by NN---** [comment]
Unfortunately this doesn't work.
It works with explicit cast, is it a bug ?

```typescript
interface QQ {
}

declare var QQ:{
    new(): QQ;
}

class QQImpl {
	constructor() {}
}

var QQ = QQImpl; // Doesn't work 

// But works with explicit cast !
var QQ = <{new():QQ;}> QQImpl;
```

**Comment by ahejlsberg** [comment]
You're getting an error because your `var QQ = QQImpl;` statement redeclares `QQ` with a (slightly) different type. The following works:
```typescript
interface QQ {
}

declare var QQ:{
    prototype: QQ; // Include prototype property to make type identical to class
    new(): QQ;
}

class QQImpl {
    constructor() {}
}

var QQ = QQImpl; // Ok, types of QQ and QQImpl identical
```
As does this:
```typescript
interface QQ {
}

declare var QQ:{
    new(): QQ;
}

class QQImpl {
    constructor() {}
}

QQ = QQImpl; // Assignment, not redeclaration
```

**Comment by NN---** [comment]
The first sample doesn't work since the QQImpl doesn't have prototype property and types are incompatible.
The second one is not what I need because it produces the following JS.
It will fail in strict parsing mode if nobody defined QQ before.

```javascript
var QQImpl = (function () {
    function QQImpl() {
    }
    return QQImpl;
})();
QQ = QQImpl; // Assignment, not redeclaration
```

**Comment by ahejlsberg** [comment]
It works in master, but apparently not with 1.3 or earlier.

I'm still not sure what you're trying to accomplish so it's hard to make further suggestions. If all you care about is declaring a class QQ that has the same shape as the interface/var QQ, then you can simply do this:
```typescript
class QQ {
}
```
This introduces ***both*** an interface QQ and a "variable" QQ (the constructor function) and in use you'll see the same effect as if you had declared the separate interface and variable.

**Comment by NN---** [comment]
Ok, the situation is the following.
I am using ES5 and browser specific features but some of them don't exist in other environments like nodejs or other browsers.
Now I want to define a variable and make everything works without getting RefereneError.
I can access to all variables through 'window' object and then it is possible to check and assign easily:
```typescript
if (window.Blob) {
   // Use window.Blob
}
```
But I don't want to change my code, moreover some definition libraries do not define global variables both globally and in window, so it means that I must rewrite definitions.

I can add pure JS file that does the simple thing: "function Blob() {}" and that's all I need.
I wanted to solve it using TypeScript language.
Since I have typing and all interfaces defined I won't make a mistake in defining variables or types.

I hope it is clear now.
Thanks.

**Comment by NN---** [comment]
Found it ! :)
Your code is totally correct but I was missing the 'var' part and it works as expected when I add it.
The only change is that I must specify the exact type of the declared variable.
I would like to see whether it can be inferred somehow there. 

```typescript
class MyBlob implements Blob {
	constructor(blobParts?: any[], options?: BlobPropertyBag) { }
	msDetachStream(): any {}
	slice(start?: number, end?: number, contentType?: string): Blob { throw new Error("Not implemented"); }
	msClose(): void {}
	type: string;
	size: number;
}
var Blob: {
	prototype: Blob;
	new(blobParts?: any[], options?: BlobPropertyBag): Blob;
} = MyBlob;
```
